jupyter_notebook/analytics.py

```diff
@@ -0,0 +1,194 @@
+
+from .base.handlers import IPythonHandler, json_errors
+from .utils import url_path_join as ujoin
+from ._version import __version__
+
+CONSENT_UNKNOWN = 0
+CONSENT_OK = 1
+CONSENT_DENIED = 2
+
+class AnalyticsSender(object):
+    _control_info = None
+    socket = None
+    enabled = False
+
+    def __init__(self, nbapp):
+        self.nbapp = nbapp
+        self.directory = osp.join(nbapp.data_dir, 'analytics')
+        self.control_file = osp.join(self.directory, 'control.json')
+        self.data_file = osp.join(nbapp.data_dir, 'analytics', 'data.json')
+        self.data = {}
+        self.ioloop = IOLoop.current()
+
+    def control_info(self):
+        if self._control_info is None:
+            try:
+                with open(self.control_file, 'r') as f:
+                    self._control_info = json.load(f)
+            except OSError as e:
+                if e.errno == errno.ENOENT:
+                    self._control_info = {}
+                else:
+                    raise
+
+        return self._control_info
+
+    def store_control_info(self):
+        assert self._control_info is not None
+        try:
+            os.makedirs(self.directory)
+        except OSError as e:
+            if e.errno != errno.EEXIST:
+                raise
+
+        with open(self.control_file, 'w') as f:
+            json.dump(self._control_info, f)
+
+    def read_data(self):
+        try:
+            with open(self.data_file, 'r') as f:
+                self.data = json.load(f)
+        except OSError as e:
+            if e.errno != errno.ENOENT:
+                raise
+
+    def store_data(self):
+        try:
+            os.makedirs(self.directory)
+        except OSError as e:
+            if e.errno != errno.EEXIST:
+                raise
+
+        with open(self.data_file, 'w') as f:
+            json.dump(self.data, f)
+
+    def consent_status(self):
+        ci = self.control_info()
+        if 'consent' in ci:
+            return CONSENT_OK if ci['consent'] else CONSENT_DENIED
+        return CONSENT_UNKNOWN
+
+    def change_consent_status(self, status):
+        ci = self.control_info()
+        if status == CONSENT_OK:
+            if not ci.get('consent', False):
+                self.enable()
+            ci['consent'] = True
+        elif status == CONSENT_DENIED:
+            ci['consent'] = False
+        else:
+            ci.pop('consent', None)
+
+        self.store_control_info()
+
+    def last_submission(self):
+        ci = self.control_info()
+        if 'last_submission' in ci:
+            return parse_date(ci['last_submission'])
+
+        return None
+
+    def startup(self):
+        webapp = self.nbapp.web_app
+
+        consent = self.consent_status()
+        if consent == CONSENT_OK:
+            self.enable()
+        elif consent == CONSENT_UNKNOWN:
+            webapp.settings['analytics_prompt'] = True
+
+        base_url = webapp.settings['base_url']
+        webapp.add_handlers(".*$", [
```

Comment:
Yeah, I have the same thoughts myself sometimes. Moving to an `service.enable(app)` pattern has a certain amount of appeal. If it is an always-on non-optional service like the others (as it is presently), it seems like it should probably be registered in a similar fashion to the others, though. The terminal one makes sense to be separate because it may not be available.